File: internal/sdkgo/conformance_test.go
package sdkgo_test

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"github.com/depot/falseflag/internal/sdkgo"
)

// TestConformance drives every tests/eval-corpus/*.json fixture through
// the SDK's snapshot-polling client and asserts that the resulting
// Decision matches the fixture's expected output byte-for-byte. This
// is the SDK-level analogue of internal/eval/cross_runtime_test.go —
// it covers the polling, parsing, and Evaluate code paths the
// evaluator-level test never touches.
func TestConformance(t *testing.T) {
	dir := corpusDir(t)
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("read corpus dir: %v", err)
	}

	type corpusFixture struct {
		ID       string          `json:"id"`
		Strategy string          `json:"strategy"`
		IR       json.RawMessage `json:"ir"`
		Context  map[string]any  `json:"context"`
		Expected struct {
			Value  any    `json:"value"`
			Reason string `json:"reason"`
			RuleID string `json:"rule_id,omitempty"`
		} `json:"expected"`
	}

	var fixtures []corpusFixture
	for _, ent := range entries {
		name := ent.Name()
		if !strings.HasSuffix(name, ".json") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(dir, name))
		if err != nil {
			t.Fatalf("read %s: %v", name, err)
		}
		var fx corpusFixture
		if err := json.Unmarshal(raw, &fx); err != nil {
			t.Fatalf("decode %s: %v", name, err)
		}
		fixtures = append(fixtures, fx)
	}

	if len(fixtures) < 25 {
		t.Fatalf("conformance corpus too small: %d fixtures (want >=25)", len(fixtures))
	}

	// Build a synthetic snapshot containing every fixture as its own
	// flag, keyed by fixture id. Then point a Client at an httptest
	// server returning that snapshot and run Evaluate per fixture.
	flagsMap := make(map[string]json.RawMessage, len(fixtures))
	for _, fx := range fixtures {
		flagsMap[fx.ID] = fx.IR
	}
	snapshotPayload, err := json.Marshal(map[string]any{
		"id":         "conformance-snapshot",
		"project_id": "conformance-project",
		"version":    1,
		"created_at": "2026-05-20T00:00:00Z",
		"compiled":   map[string]any{"flags": flagsMap},
	})
	if err != nil {
		t.Fatalf("marshal snapshot: %v", err)
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !strings.HasSuffix(r.URL.Path, "/snapshots/latest") {
			http.NotFound(w, r)
			return
		}
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write(snapshotPayload)
	}))
	defer srv.Close()

	c, err := sdkgo.NewClient(sdkgo.Options{
		BaseURL:      srv.URL,
		ProjectSlug:  "conformance",
		PollInterval: -1,
		Logger:       slog.New(slog.NewTextHandler(io.Discard, nil)),
	})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	if err := c.Start(context.Background()); err != nil {
		t.Fatalf("Start: %v", err)
	}
	defer c.Stop()

	if c.Snapshot() == nil {
		t.Fatal("snapshot nil after Start")
	}

	matched := 0
	for _, fx := range fixtures {
		t.Run(fx.ID, func(t *testing.T) {
			got := c.Evaluate(fx.ID, fx.Context)

			gotVal, _ := json.Marshal(got.Value)
			wantVal, _ := json.Marshal(fx.Expected.Value)
			if string(gotVal) != string(wantVal) {
				t.Errorf("value: got %s, want %s", gotVal, wantVal)
			}
			if got.Reason != fx.Expected.Reason {
				t.Errorf("reason: got %q, want %q", got.Reason, fx.Expected.Reason)
			}
			if fx.Expected.RuleID != "" && got.RuleID != fx.Expected.RuleID {
				t.Errorf("rule_id: got %q, want %q", got.RuleID, fx.Expected.RuleID)
			}
		})
		matched++
	}
	fmt.Printf("conformance: %d/%d fixtures match\n", matched, len(fixtures))
}

func corpusDir(t *testing.T) string {
	t.Helper()
	_, here, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	return filepath.Join(filepath.Dir(here), "..", "..", "tests", "eval-corpus")
}

File: js/packages/sdk-js/tests/conformance.test.ts
import { readFileSync, readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { createClient } from "../src/client.js";
import type {
  Decision,
  DecisionReason,
  RulesTree,
  Strategy,
} from "../src/index.js";

interface Fixture {
  id: string;
  description: string;
  strategy: Strategy;
  ir: RulesTree;
  context: Record<string, unknown>;
  expected: {
    value: unknown;
    reason: DecisionReason;
    rule_id?: string;
  };
}

function corpusDir(): string {
  const here = dirname(fileURLToPath(import.meta.url));
  // js/packages/sdk-js/tests/ → repo root is 5 levels up.
  return join(here, "..", "..", "..", "..", "tests", "eval-corpus");
}

function loadCorpus(): Fixture[] {
  const dir = corpusDir();
  return readdirSync(dir)
    .filter((f) => f.endsWith(".json"))
    .sort()
    .map((f) => JSON.parse(readFileSync(join(dir, f), "utf8")) as Fixture);
}

describe("SDK conformance against tests/eval-corpus", () => {
  const fixtures = loadCorpus();

  it("loads at least 25 fixtures", () => {
    expect(fixtures.length).toBeGreaterThanOrEqual(25);
  });

  it("evaluates every fixture identically to the expected Decision", async () => {
    // Build a synthetic snapshot containing every fixture keyed by id.
    const flagsMap: Record<string, RulesTree> = {};
    for (const fx of fixtures) flagsMap[fx.id] = fx.ir;

    const snapshotBody = {
      id: "conformance-snapshot",
      project_id: "conformance-project",
      version: 1,
      created_at: "2026-05-20T00:00:00Z",
      compiled: { flags: flagsMap },
    };

    const fakeFetch: typeof fetch = async () =>
      new Response(JSON.stringify(snapshotBody), {
        status: 200,
        headers: { "content-type": "application/json" },
      });

    const client = createClient({
      baseUrl: "http://api",
      projectSlug: "conformance",
      fetch: fakeFetch,
      pollIntervalMs: 0,
      warn: () => {},
    });
    await client.start();
    expect(client.getSnapshot()).not.toBeNull();

    for (const fx of fixtures) {
      const got: Decision = await client.evaluate(fx.id, fx.context);
      expect(JSON.stringify(got.value), `${fx.id} value`).toBe(
        JSON.stringify(fx.expected.value),
      );
      expect(got.reason, `${fx.id} reason`).toBe(fx.expected.reason);
      if (fx.expected.rule_id) {
        expect(got.rule_id, `${fx.id} rule_id`).toBe(fx.expected.rule_id);
      }
    }
    client.stop();
  });
});
